source/server/listener_impl.h

```diff
@@ -60,11 +60,16 @@ class ListenSocketFactoryImpl : public Network::ListenSocketFactory,
    * @return the socket shared by worker threads; otherwise return null.
    */
   Network::SocketOptRef sharedSocket() const override {
+    // If a tcp listener socket doesn't bind to port, there is no listen socket so there is no
+    // shared socket.
+    if (socketType() == Network::Socket::Type::Stream && !bind_to_port_) {
```

Comment:
Does this apply only to Stream sockets or should it also apply to other socket types?

Reply:
That's a good question!
I think the sharedSocket() matters only if my imagined `bind_to_port` is on.
It seems udp listener's bind_to_port has a different semantic 
@danzh2010 Could you add some comment on the bind_to_port usage in udp/quic listener?

Reply:
Regardless the semantic of bind_to_port in UDP world, the implementation of UDP on bind_to_port = false is that a socket is created so sharedSocket() must returns it.

Reply:
So the answer to the original question:
As per the current impl of udp listener, sharedSocket() should return it as in this PR.

However, if udp listener turns to not create socket on bind_to_port==false, this sharedSocket() need to align with that change.

Reply:
It would be good to have the behaviors in sync.  bind_to_port == false doesn't make much sense to me for UDP, I wonder if it should be disallowed.

Reply:
bind_to_port works the same for UDP and TCP from reading the NetworkListenSocket construction code. And I think we always need to bind the listening socket. What are the not binding cases?

That being said, I'm wondering if returning nullopt in sharedSocket() based on bind_to_port is the right approach? According to the contract sharedSocket() returns nullopt when the sockets among worker threads are not shared, so it is called in ListenerManagerImpl::drainListener() to decide whether to close shared socket or not. This PR seems to change that contract. Should we also change the corresponding logic of closing socket?

Reply:
> bind_to_port works the same for UDP and TCP from reading the NetworkListenSocket construction code. And I think we always need to bind the listening socket. What are the not binding cases?

There is a logic that tcp packet headers can be mangled in kernel, in win and linux.
Envoy supports this scenario by figuring out the address before mangling, and hand off the tcp connection to the tcp listener that doesn't physically listen on that port but mark the listener as accepting these mangled connections.

Technically the udp packet could be mangled, but my perception is that UDP listener in Envoy does not (yet) have the ability to recover or redirect. That makes "bind_to_port == false" udp listener not reasonable: such udp listener can be somehow created but not referenced after. I am wondering if we can reject such config with runtime flag.

Reply:
> That being said, I'm wondering if returning nullopt in sharedSocket() based on bind_to_port is the right approach? According to the contract sharedSocket() returns nullopt when the sockets among worker threads are not shared, so it is called in ListenerManagerImpl::drainListener() to decide whether to close shared socket or not. This PR seems to change that contract. Should we also change the corresponding logic of closing socket?

I agree this PR is somewhat changing the contract of sharedSocket(). Essentially we have two approaches, one is to provide a dedicate socket impl for listener doesn't bind to port. That specialized socket can be returned by sharedSocket(). The other approach in this PR is not to allow sharedSocket() from the beginning. Close is not an issue as per implementation: no underlying BSD socket is created so we are not leaking fd.

Reply:
As I mentioned elsewhere, I'm currently working on simplifying all of this code, so I would prefer to not make major changes here which I'm going to undo anyway. Per my above comment, why can't we just check if io_handle_ is null in various places?

Reply:
> I would prefer to not make major changes here which I'm going to undo anyway. 

I feel this PR has two part: sharedSocket() and the the ListenerSocketImpl.

Your changes will delete sharedSocket(). However, the latter ListenerSocketImpl is needed. The reason is that your commits seems create number of concurrency listen socket regardless it's bind_to_port, for simplicity. I don't know how is that listenersocket closed correctly... I could be wrong though.

Reply:
> Your changes will delete sharedSocket(). However, the latter ListenerSocketImpl is needed. The reason is that your commits seems create number of concurrency listen socket regardless it's bind_to_port, for simplicity. I don't know how is that listenersocket closed correctly... I could be wrong though.

Sorry I can't parse this. We can review my other change when I post it.

Reply:
Oh, i had the wrong impression of udp listen socket always need an io_handle even when bind_to_port is false. I discovered this "fact" and I now incline to believe the fact is b/c a test case is improperly using listener socket.